File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/visual/DItemTree.java
/*
 * Copyright (c) 1990-2018 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DItemTree.java 35325 2018-08-29 11:27:17Z iayadi $
 */

package org.kopi.vkopi.lib.ui.swing.visual;

import java.awt.BorderLayout;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.List;

import javax.swing.JOptionPane;
import javax.swing.JScrollPane;
import javax.swing.JTree;
import javax.swing.UIManager;
import javax.swing.event.TreeExpansionEvent;
import javax.swing.event.TreeExpansionListener;
import javax.swing.event.TreeSelectionEvent;
import javax.swing.event.TreeSelectionListener;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreeNode;
import javax.swing.tree.TreePath;

import org.kopi.galite.util.base.Utils;
import org.kopi.galite.visual.Item;
import org.kopi.galite.visual.Action;
import org.kopi.galite.visual.MessageCode;
import org.kopi.galite.visual.UItemTree;
import org.kopi.galite.visual.VException;
import org.kopi.galite.visual.VExecFailedException;
import org.kopi.galite.visual.VItemTree;
import org.kopi.galite.visual.VlibProperties;

public class DItemTree extends DWindow implements UItemTree {

  // --------------------------------------------------------------------
  // CONSTRUCTOR
  // --------------------------------------------------------------------

  @SuppressWarnings("deprecation")
  public DItemTree(VItemTree model) {
    super(model);
    tree = new Tree(model.getRoot());

    tree.addMouseListener(new MouseAdapter() {
      private long lastClick;

      public void mouseClicked(MouseEvent e) {
        if (e.getClickCount() == 2 && getModel().isNoEdit()) {
          changeSelectionState();
        } else {
          if (e.getWhen() - lastClick < 400) {
            // for slow NT users
            changeSelectionState();
          }
        }
        lastClick = e.getWhen();
      }
    });

    tree.addKeyListener(new KeyAdapter() {
      public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_SPACE
            || (e.getKeyCode() == KeyEvent.VK_ENTER
            && getSelectedNode().isLeaf()))
        {
          e.consume();
          changeSelectionState();
        }
      }
    });

    tree.addTreeExpansionListener(new TreeExpansionListener() {

      public void treeExpanded(TreeExpansionEvent event) {
        setTree();
      }

      public void treeCollapsed(TreeExpansionEvent event) {
        setTree();
      }
    });

    tree.getSelectionModel().addTreeSelectionListener(new TreeSelectionListener() {
      public void valueChanged(TreeSelectionEvent e) {
        setTree();
      }
    });

    tree.setCellRenderer(new ItemRenderer(getModel().isNoEdit(), getModel().isSingleSelection(), getModel().isLocalised()));
    tree.putClientProperty("JTree.lineStyle", "None");

    /* Make tree ask for the height of each row. */
    tree.setRowHeight(-1);
    tree.setBackground(UIManager.getColor("menu.background"));

    JScrollPane sp = new JScrollPane();

    sp.setBorder(null);
    sp.getViewport().add(tree);
    getContentPanel().setLayout(new BorderLayout());
    getContentPanel().add(sp, BorderLayout.CENTER);

    if (tree.getRowCount() > 0) {
      tree.setSelectionInterval(0, 0);
    }
    setTree();
    tree.requestFocusInWindow();
  }

  // --------------------------------------------------------------------
  // IMPLEMENTATIONS
  // --------------------------------------------------------------------

  public void setTree() {
    DefaultMutableTreeNode                      node;
    boolean                                     rootItem;

    if (getModel() != null) {   // The model can be destroyed by it's itemTreeManager:
                                // the save action close the Item Tree window
      node = getSelectedNode();
      getModel().setActorEnabled(VItemTree.CMD_QUIT, true);
      if (node != null) {
        rootItem = ((Item)node.getUserObject()).getId() == -1;
        getModel().setActorEnabled(VItemTree.CMD_ADD, getModel().isInsertMode());
        getModel().setActorEnabled(VItemTree.CMD_REMOVE, getModel().isInsertMode() && !rootItem);
        getModel().setActorEnabled(VItemTree.CMD_EDIT, getModel().isInsertMode() && !rootItem);
        getModel().setActorEnabled(VItemTree.CMD_LOCALISE, (getModel().isLocalised() && getModel().isInsertMode() && !rootItem));
        getModel().setActorEnabled(VItemTree.CMD_SELECT, !getModel().isNoEdit() && !rootItem);
        getModel().setActorEnabled(VItemTree.CMD_DEFAULT, getModel().isMultiSelectionDefaultValue() && !rootItem);
        if (node.isLeaf()) {
          getModel().setActorEnabled(VItemTree.CMD_FOLD, false);
          getModel().setActorEnabled(VItemTree.CMD_UNFOLD, false);
        } else {
          if (tree.isExpanded(tree.getSelectionPath())) {
            getModel().setActorEnabled(VItemTree.CMD_FOLD, true);
            getModel().setActorEnabled(VItemTree.CMD_UNFOLD, false);
          } else {
            getModel().setActorEnabled(VItemTree.CMD_FOLD, false);
            getModel().setActorEnabled(VItemTree.CMD_UNFOLD, true);
          }
        }
      } else {
        getModel().setActorEnabled(VItemTree.CMD_FOLD, false);
        getModel().setActorEnabled(VItemTree.CMD_UNFOLD, false);
        getModel().setActorEnabled(VItemTree.CMD_ADD, false);
        getModel().setActorEnabled(VItemTree.CMD_REMOVE, false);
        getModel().setActorEnabled(VItemTree.CMD_EDIT, false);
        getModel().setActorEnabled(VItemTree.CMD_LOCALISE, false);
        getModel().setActorEnabled(VItemTree.CMD_SELECT, false);
        getModel().setActorEnabled(VItemTree.CMD_DEFAULT, false);
      }
      getModel().setActorEnabled(VItemTree.CMD_SAVE, getModel().isChanged());
    }
  }

  public void run() throws VException {
    setVisible(true);
  }

  /**
   * Returns the TreeNode instance that is selected in the tree.
   * If nothing is selected, null is returned.
   */
  protected DefaultMutableTreeNode getSelectedNode() {
    TreePath    selPath = tree.getSelectionPath();

    if (selPath != null) {
      return (DefaultMutableTreeNode)selPath.getLastPathComponent();
    } else {
      return null;
    }
  }

  /**
   * Sets item selection state
   */
  public void setSelectedItem() {
    DefaultMutableTreeNode              selectedNode;
    Item                                selectedItem;
    DefaultTreeModel                    treeModel;
    DefaultMutableTreeNode              rootNode;

    selectedNode = getSelectedNode();
    if (selectedNode != null) {
      selectedItem = (Item)selectedNode.getUserObject();

      treeModel = ((DefaultTreeModel)tree.getModel());
      if (getModel().isSingleSelection()) {
        rootNode = (DefaultMutableTreeNode)treeModel.getRoot();
        if (!selectedItem.getSelected()) {
          unselectAll(treeModel, rootNode);
        }
      }
      selectedItem.setSelected(!selectedItem.getSelected());
      treeModel.nodeChanged(selectedNode);
      getModel().refresh();
    }
  }

  /**
   * Set selected item as default element
   */
  public void setDefaultItem() {
    DefaultMutableTreeNode              selectedNode;
    Item                                selectedItem;
    DefaultTreeModel                    treeModel;
    DefaultMutableTreeNode              rootNode;

    selectedNode = getSelectedNode();

    if (selectedNode != null) {
      treeModel = ((DefaultTreeModel)tree.getModel());
      rootNode = (DefaultMutableTreeNode)treeModel.getRoot();
      selectedItem = (Item)selectedNode.getUserObject();

      if (!selectedItem.getDefaultItem()) {
        setDefault(treeModel, rootNode);
      }
      selectedItem.setDefaultItem(!selectedItem.getDefaultItem());
      selectedItem.setSelected(true);
      treeModel.nodeChanged(selectedNode);
      getModel().refresh();
    }
  }

  /**
   * Set default value of all children
   */
  private void setDefault(DefaultTreeModel treeModel,
                          DefaultMutableTreeNode node)
  {
    for (int i = 0; i < node.getChildCount(); i++) {
      DefaultMutableTreeNode child = (DefaultMutableTreeNode)node.getChildAt(i);
      Item item = (Item)child.getUserObject();

      if (item.getDefaultItem()) {
        item.setDefaultItem(false);
        treeModel.nodeChanged(child);
      }
      if (child.getChildCount() > 0) {
        setDefault (treeModel, child);
      }
    }
  }

  /**
   * Set selection value of all children
   */
  private void unselectAll(DefaultTreeModel treeModel,
                           DefaultMutableTreeNode node)
  {
    for (int i = 0; i < node.getChildCount(); i++) {
      DefaultMutableTreeNode child = (DefaultMutableTreeNode)node.getChildAt(i);
      Item item = (Item)child.getUserObject();

      if (item.getSelected()) {
        item.setSelected(false);
        treeModel.nodeChanged(child);
      }
      if (child.getChildCount() > 0) {
        unselectAll(treeModel, child);
      }
    }
  }

  /**
   * Change item selection state
   */
  private void changeSelectionState() {
    getModel().performAsyncAction(new Action("change_selection_state") {
      public void execute() {
        if (!getModel().isNoEdit()) {
          setSelectedItem();
        }
      }
    });
  }

  /**
   * Insert new item
   */
  public void addItem() throws VException {
    DefaultMutableTreeNode              selectedNode;
    DefaultTreeModel                    treeModel;
    DefaultMutableTreeNode              newNode;
    String                              newItem;
    int                                 selectedRow;
    int                                 maxLength;
    Item                                parent;
    Item                                item;

    maxLength = Math.min(getModel().MAX_LENGTH, getModel().getNameLength());
    selectedRow = tree.getSelectionRow();
    selectedNode = getSelectedNode();
    if (selectedNode != null) {
      parent = (Item)selectedNode.getUserObject();
      if (getModel().getDepth() > 0 && ((parent.getLevel() + 1) > getModel().getDepth())) {
        throw new VExecFailedException(MessageCode.Companion.getMessage("VIS-00069" , getModel().getDepth()));
      }

      do {
        newItem = JOptionPane.showInputDialog(this,
                                              MessageCode.Companion.getMessage("VIS-00068", maxLength),
                                              VlibProperties.getString("New_item"),
                                              JOptionPane.PLAIN_MESSAGE);
      } while (newItem != null && newItem.length() > maxLength);
      if (!(newItem == null || newItem.trim().length() == 0)) {
        newItem = newItem.trim();
        treeModel = ((DefaultTreeModel)tree.getModel());
        item = new Item(getModel().getNextId(),
                        ((Item)selectedNode.getUserObject()).getId(),
                        newItem,
                        null,
                        null,
                        false,
                        false,
                        null,
                        newItem);
        item.setLevel(parent.getLevel() + 1);
        newNode = new DefaultMutableTreeNode(item);
        selectedNode.add(newNode);
        if(selectedNode.getChildCount() > 0) {
          treeModel.reload(selectedNode);
        } else {
          treeModel.nodeChanged(selectedNode);
        }
        if (selectedRow > 0) {
          tree.expandRow(selectedRow);
        }
      }
    }
  }

  /**
   * Remove selected item from the tree
   */
  public void removeSelectedItem() {
    DefaultMutableTreeNode              selectedNode;
    DefaultTreeModel                    treeModel;

    selectedNode = getSelectedNode();
    treeModel = ((DefaultTreeModel)tree.getModel());
    if (selectedNode != null) {
      treeModel = ((DefaultTreeModel)tree.getModel());
      DefaultMutableTreeNode rootNode = (DefaultMutableTreeNode)treeModel.getRoot();

      if (!getModel().isRemoveDescendantsAllowed()) {
        attacheToParent(treeModel, selectedNode, (DefaultMutableTreeNode)selectedNode.getParent());
      }
      removeItem(treeModel, rootNode, (Item)selectedNode.getUserObject());
    }
  }

  /**
   * Attache children to removed item parent
   */
  public void attacheToParent(DefaultTreeModel treeModel,
                              DefaultMutableTreeNode node,
                              DefaultMutableTreeNode parent)
  {
    for (int i = 0; i < node.getChildCount(); i++) {
      DefaultMutableTreeNode    child = (DefaultMutableTreeNode)node.getChildAt(i);
      Item                      childItem = (Item)child.getUserObject();

      childItem.setParent(((Item)parent.getUserObject()).getId());
      parent.add(child);
      if(parent.getChildCount() > 0){
        treeModel.reload(parent);
      } else {
        treeModel.nodeChanged(parent);
      }
      setLevel(treeModel, child);
    }
  }

  /**
   * Sets level of all children of an item
   */
  public void setLevel(DefaultTreeModel treeModel,
                       DefaultMutableTreeNode node)
  {
    for (int i = 0; i < node.getChildCount(); i++) {
      DefaultMutableTreeNode child = (DefaultMutableTreeNode)node.getChildAt(i);

      if (child.getChildCount() > 0) {
        setLevel(treeModel, child);
      }
    }
    ((Item)node.getUserObject()).decrementLevel();
  }

  /**
   * Remove an item from the tree
   */
  private void removeItem(DefaultTreeModel treeModel,
                          DefaultMutableTreeNode node,
                          Item itemToRemove)
  {
    for (int i = 0; i < node.getChildCount(); i++) {
      DefaultMutableTreeNode child = (DefaultMutableTreeNode)node.getChildAt(i);
      Item childItem = (Item)child.getUserObject();
      if (childItem.getId() == itemToRemove.getId()) {
        getModel().getRemovedItems().add(itemToRemove);
        node.remove(i);
        treeModel.reload(node);
        return;
      } else {
        if (child.getChildCount() > 0) {
          removeItem(treeModel, child, itemToRemove);
        }
      }
    }
  }

  /**
   * Edit the selected item
   */
  public void editSelectedItem() {
    DefaultMutableTreeNode              selectedNode;
    Item                                selectedItem;
    DefaultTreeModel                    treeModel;
    String                              newName;
    int                                 maxLength;

    maxLength = Math.min(getModel().MAX_LENGTH, getModel().getNameLength());
    selectedNode = getSelectedNode();

    if (selectedNode != null) {
      selectedItem = (Item)selectedNode.getUserObject();
      do {
        newName = (String) JOptionPane.showInputDialog(this,
                                                       MessageCode.Companion.getMessage("VIS-00068", maxLength),
                                                       VlibProperties.getString("OpenLine"),
                                                       JOptionPane.PLAIN_MESSAGE,
                                                       null,
                                                       null,
                                                       selectedItem.getName());
      } while (newName != null && newName.length() > maxLength);
      if (!(newName == null || newName.trim().length() == 0)) {
        treeModel = ((DefaultTreeModel)tree.getModel());
        selectedItem.setName(newName);
        treeModel.nodeChanged(selectedNode);
      }
    }
  }

  /**
   * Localise the selected item
   */
  public void localiseSelectedItem() {
    DefaultMutableTreeNode              selectedNode;
    Item                                selectedItem;
    DefaultTreeModel                    treeModel;
    String                              localisedName;
    int                                 maxLength;

    maxLength = Math.min(getModel().MAX_LENGTH, getModel().getLocalisedNameLength());
    selectedNode = getSelectedNode();

    if (selectedNode != null) {
      selectedItem = (Item)selectedNode.getUserObject();
      do {
        localisedName = (String)JOptionPane.showInputDialog(this,
                                                            MessageCode.Companion.getMessage("VIS-00068", maxLength),
                                                            VlibProperties.getString("OpenLine"),
                                                            JOptionPane.PLAIN_MESSAGE,
                                                            null,
                                                            null,
                                                            selectedItem.getLocalisedName());
      } while (localisedName != null && localisedName.length() > maxLength);
      if (!(localisedName == null || localisedName.trim().length() == 0)) {
        treeModel = ((DefaultTreeModel)tree.getModel());
        selectedItem.setLocalisedName(localisedName);
        treeModel.nodeChanged(selectedNode);
      }
    }
  }

  // --------------------------------------------------------------------
  // ACCESSORS
  // --------------------------------------------------------------------

  public UTreeComponent getTree() {
    return tree;
  }

  public VItemTree getModel() {
    return (VItemTree) super.getModel();
  }

  //------------------------------------------------------------
  // UTREE IMPLEMENTATION
  //------------------------------------------------------------

  public class Tree extends JTree implements UTreeComponent {

    //--------------------------------------------------------
    // CONSTRUCTOR
    //--------------------------------------------------------

    public Tree(TreeNode root) {
      super(root);
      this.root = root;
    }

    public boolean isExpanded(Object path) {
      return super.isExpanded((TreePath) path);
    }

    public boolean isCollapsed(Object path) {
      return super.isCollapsed((TreePath) path);
    }

    public int getSelectionRow() {
      return super.getSelectionRows()[0];
    }

    /**
     * Return items array
     */
    public Item[] getItems() {
      List<Item> itemsList = getItems(root);

      if (itemsList != null && itemsList.size() > 0) {
        return (Item[])Utils.toArray(itemsList, Item.class);
      }

      return null;
    }

    /**
     * Return items list
     */
    public List<Item> getItems(TreeNode node) {
      List<Item>                items;
      Item                      item;

      items = new ArrayList<Item>();
      item = (Item)((DefaultMutableTreeNode)node).getUserObject();
      item.setChildCount(node.getChildCount());
      if (item.getId() >= 0) {
        items.add(item);
      }

      if (node.getChildCount() > 0) {
        for (int i = 0; i < node.getChildCount(); i++) {
          items.addAll(getItems(node.getChildAt(i)));
        }
      }

      return items;
    }

    /**
     * Return items array
     */
    @Override
    public Item getRootItem() {
      return getRootItem(root);
    }

    /**
     * Return items list
     */
    public Item getRootItem(TreeNode node) {
      Item[]            children;
      Item              item;

      item = (Item)((DefaultMutableTreeNode)node).getUserObject();
      item.setChildCount(node.getChildCount());

      if (node.getChildCount() > 0) {
        children = new Item[node.getChildCount()];
        for (int i = 0; i < node.getChildCount(); i++) {
          children[i] = getRootItem(node.getChildAt(i));
        }
        item.setChildren(children);
      }

      return item;
    }

    /**
     * Return true if the item name done is unique in this tree
     */
    public boolean isUnique(String name) {
      // TODO: Should be implemented when the item name
      //       may be not unique in the tree
      return true;
    }

    /**
     * Ensures that the selected nodes are expanded and viewable.
     */
    public void expandTree() {
    }

    //------------------------------------------------------------
    // DATA MEMBERS
    //------------------------------------------------------------

    TreeNode                            root;

    private static final long           serialVersionUID = -3954942255139034033L;
  }

  // --------------------------------------------------------------------
  // DATA MEMBERS
  // --------------------------------------------------------------------

  private Tree                          tree;

  private static final long             serialVersionUID = 8814499549529628999L;
}

File: galite-core/src/main/kotlin/org/kopi/galite/ui/vaadin/visual/DItemTree.kt
/*
 * Copyright (c) 2013-2021 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.galite.ui.vaadin.visual

import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.TreeNode

import org.kopi.galite.ui.vaadin.base.BackgroundThreadHandler.access
import org.kopi.galite.ui.vaadin.base.BackgroundThreadHandler.accessAndPush
import org.kopi.galite.ui.vaadin.base.Styles
import org.kopi.galite.ui.vaadin.base.Utils.findMainWindow
import org.kopi.galite.ui.vaadin.window.PopupWindow
import org.kopi.galite.visual.Item
import org.kopi.galite.visual.MessageCode
import org.kopi.galite.visual.UItemTree
import org.kopi.galite.visual.VException
import org.kopi.galite.visual.VExecFailedException
import org.kopi.galite.visual.VItemTree
import org.kopi.galite.visual.VRuntimeException
import org.kopi.galite.visual.VlibProperties

import com.vaadin.flow.component.ComponentEventListener
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.grid.ItemClickEvent
import com.vaadin.flow.component.html.Label
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.textfield.TextField

/**
 * The `DItemTree` is the vaadin implementation of the
 * [UItemTree].
 *
 *
 * The implementation is based on [DWindow]
 *
 * @param model The item tree model.
 *
 * TODO Externalize favorites handling.
 */
class DItemTree(model: VItemTree) : DWindow(model), UItemTree {

  // --------------------------------------------------
  // DATA MEMBERS
  // --------------------------------------------------
  private var tree = Tree(model.root, model.isNoEdit(), model.isLocalised)
  private var inputDialog: PopupWindow? = null
  private lateinit var editTextField: TextField
  private var errorLabel: Label? = null
  private var newItem = false
  private var localisation = false

  // --------------------------------------------------
  // CONSTRUCTOR
  // --------------------------------------------------
  init {
    val content = VerticalLayout(tree)
    tree.addItemClickListener(ItemClickHandler())
    //tree.addCollapseListener(DItemTree.CollapseHandler())
    //tree.addExpandListener(DItemTree.ExpandHandler())
    content.className = "itemtree"
    /*tree.addItemSetChangeListener(object : ItemSetChangeListener() { TODO
      fun containerItemSetChange(event: ItemSetChangeEvent?) {
        setTree()
      }
    })
    tree.addValueChangeListener(object : ValueChangeListener() {
      fun valueChange(event: ValueChangeEvent) {
        val itemId: Any = event.getProperty().getValue() ?: return
        val item = tree.getITEM(itemId)
        tree.setIcon(item, isLeaf())
        setTree()
      }
    })*/
    model.setDisplay(this)
    tree.setSizeUndefined()
    if (model.isNoEdit()) {
      tree.expandRow(0)
    } else {
      tree.expandTree()
    }
    setTree()
    // allow scrolling when an overflow is detected
    content.width = "900px"
    content.height = "800px"
    setContent(content)
  }

  // --------------------------------------------------------------------
  // IMPLEMENTATIONS
  // --------------------------------------------------------------------
  /**
   * Sets item selection state.
   */
  override fun setSelectedItem() {
    val item = getSelectedItem()
    if (item != null) {
      if (!getModel()!!.isNoEdit()) {
        if (getModel()!!.isSingleSelection()) {
          if (!item.isSelected) {
            unselectAll(-1)
            item.isSelected = true
          } else {
            item.isSelected = false
          }
        } else {
          item.isSelected = !item.isSelected
        }
        // tree.setIcon(item) TODO
        getModel()!!.refresh()
      } else {
        if (tree.isExpanded(tree.selectedItem)) {
          tree.collapse(tree.selectedItem)
        } else {
          tree.expand(tree.selectedItem)
        }
      }
    }
  }

  /**
   * Set selection value to false for all children
   */
  private fun unselectAll(itemID: Int) {
    val children = tree.getChildrenOf(itemID)
    children?.forEach {
      val item = tree.getITEM(it) as Item
      if (item.isSelected) {
        item.isSelected = false
        // tree.setIcon(item) TODO
      }
      val Grandsons = tree.getChildrenOf(item.id)
      if (Grandsons != null && Grandsons.isNotEmpty()) {
        unselectAll(item.id)
      }
    }
  }

  override fun setDefaultItem() {
    val item = getSelectedItem()
    if (item != null) {
      if (!item.isDefaultItem) {
        setDefault(-1)
        item.isDefaultItem = true
      } else {
        item.isDefaultItem = false
      }
      item.isSelected = true
      // tree.setIcon(item) TODO
      getModel()!!.refresh()
    }
  }

  /**
   * Set default value for all children
   */
  private fun setDefault(itemID: Int) {
    val children = tree.getChildrenOf(itemID)
    children?.forEach {
      val item = tree.getITEM(it) as Item
      if (item.isDefaultItem) {
        item.isDefaultItem = false
        // tree.setIcon(item) TODO
      }
      val Grandsons = tree.getChildrenOf(item.id)
      if (Grandsons != null && Grandsons.isNotEmpty()) {
        setDefault(item.id)
      }
    }
  }

  override fun addItem() {
    val item = getSelectedItem()
    if (item != null) {
      if (getModel()!!.depth > 0 && item.level + 1 > getModel()!!.depth) {
        throw VExecFailedException(MessageCode.getMessage("VIS-00069", getModel()!!.depth))
      }
      localisation = false
      showInputDialog(item, true)
    }
  }

  override fun removeSelectedItem() {
    val item = getSelectedItem()
    if (item != null) {
      access(currentUI) {
        if (getModel()!!.isRemoveDescendantsAllowed) {
          removeChildren(item)
        } else {
          attacheToParent(item)
        }
        getModel()!!.removedItems.add(item)
        tree.treeData.removeItem(tree.getItem(item.id))
        tree.valueChanged()
      }
    }
  }

  /**
   * Remove all children of an item
   * @param parentId the parent ID
   */
  fun removeChildren(parent: Item) {
    val children = tree.getChildrenOf(parent.id)
    children?.forEach {
      removeChildren(tree.getITEM(it) as Item)
    }
    getModel()!!.removedItems.add(parent)
    tree.treeData.removeItem(tree.getItem(parent.id))
  }

  /**
   * Attache children to removed item parent
   */
  fun attacheToParent(item: Item) {
    val children = tree.getChildrenOf(item.id)
    children?.forEach {
      tree.treeData.setParent(it, tree.getItem(item.parent))
      tree.getITEM(it)!!.parent = item.parent
      setLevel(it)
    }
  }

  /**
   * Sets level value for all children of an item
   * @param itemId the item ID
   */
  fun setLevel(itemId: TreeNode?) {
    val children = tree.getChildrenOf(itemId)
    children?.forEach {
      setLevel(it)
    }
    tree.getITEM(itemId)!!.decrementLevel()
  }

  /**
   * Localise the selected item
   */
  override fun localiseSelectedItem() {
    val item = getSelectedItem()
    if (item != null) {
      localisation = true
      showInputDialog(item, false)
    }
  }

  /**
   * Edit selected item's name
   */
  override fun editSelectedItem() {
    val item = getSelectedItem()
    if (item != null) {
      localisation = false
      showInputDialog(item, false)
    }
  }

  /**
   * Shows a modal window in an inputDialog view. This will handle
   * a window view
   * @param item item to edit or add new child
   * @param newItem if true, it is a new item to insert
   */
  protected fun showInputDialog(
    item: Item,
    newItem: Boolean
  ) {
    val maxLength =
      VItemTree.MAX_LENGTH.coerceAtMost(if (localisation) getModel()!!.localisedNameLength else getModel()!!.nameLength)
    if (inputDialog == null) {
      createInputDialog(localisation)
    }
    accessAndPush(currentUI) {
      editTextField.value = if (newItem) "" else if (localisation) if (item.localisedName != null) item.localisedName else "" else item.name
      editTextField.maxLength = maxLength
      editTextField.width = "" + maxLength + "em"
      this.newItem = newItem
      inputDialog!!.setCaption(if (localisation) item.name else VlibProperties.getString(if (newItem) "New_item" else "OpenLine"))
      editTextField.focus()
      inputDialog!!.open()
    }
  }

  /**
   * Create an input dialog
   *
   * @param localisation if true, the edit text will
   * contain the localised item's name
   */
  private fun createInputDialog(localisation: Boolean) {
    val popupContent = VerticalLayout()
    popupContent.width = "400px"
    editTextField = TextField("")
    //editTextField.setWidth(180, Unit.PIXELS);
    editTextField.className = Styles.TEXT_INPUT
    errorLabel = Label(MessageCode.getMessage("VIS-00020", getModel()!!.getTitle()))
    errorLabel!!.isVisible = false
    errorLabel!!.addClassName("notificationlabel")
    popupContent.add(errorLabel)
    popupContent.add(editTextField)
    // popupContent.setComponentAlignment(editTextField, Alignment.MIDDLE_CENTER) TODO
    val buttonsContainer = HorizontalLayout()
    val okButton = Button("OK") {
      editItem(getSelectedItem())
    }
    okButton.width = "80px"
    okButton.className = "inputbutton"
    val cancelButton = Button("Annuler") {
      errorLabel!!.isVisible = false
      inputDialog!!.close()
    }
    cancelButton.width = "80px"
    cancelButton.className = "inputbutton"
    buttonsContainer.add(okButton)
    buttonsContainer.add(cancelButton)
    popupContent.add(buttonsContainer)
    // popupContent.setComponentAlignment(buttonsContainer, Alignment.MIDDLE_CENTER) TODO
    inputDialog = PopupWindow(findMainWindow())
    inputDialog!!.isModal = true
    inputDialog!!.setContent(popupContent)
    inputDialog!!.className = "inputdialog"
  }

  /**
   * Edit or add new item to the tree
   *
   * @param item        the selected item
   */
  private fun editItem(item: Item?) {
    val itemName: String = editTextField.value
    if (itemName.isNotEmpty()) {
      if (getTree().isUnique(itemName)) {
        errorLabel!!.isVisible = false
        inputDialog!!.close()
        if (newItem) {
          addItem(itemName, null, null, item)
        } else {
          if (localisation) {
            item!!.localisedName = itemName
          } else {
            item!!.name = itemName
          }
          /*tree.getItem(item.id).getItemProperty(Tree.ITEM_PROPERTY_NAME).setValue( TODO
            item.getFormattedName(
              getModel()!!.isLocalised
            )
          )*/
        }
        tree.valueChanged()
      } else {
        errorLabel!!.isVisible = true
        editTextField.focus()
      }
    }
  }

  /**
   * Add new Item
   *
   * @param itemName            the new item's name
   * @param localisedName       the itme localised name
   * @param description         the itme description
   * @param parent              the parent item
   */
  private fun addItem(
    itemName: String,
    localisedName: String?,
    description: String?,
    parent: Item?
  ) {
    val node = DefaultMutableTreeNode()
    val item = Item(
      getModel()!!.getNextId(),
      parent!!.id,
      itemName,
      localisedName,
      description,
      false,
      false,
      null,
      itemName
    )
    node.userObject = item
    val nodeComponent = tree.addItemComponent(node)

    item.level = parent.level + 1
    // dataItem.getItemProperty(Tree.ITEM_PROPERTY_ICON).setValue(item.icon) TODO
    tree.treeData.addItem(tree.getItem(parent.id), tree.getItem(item.id))
    nodeComponent.setIcon(item)
    tree.expand(tree.getItem(parent.id))
  }

  override fun setTree() {
    val item: Item?
    if (getModel() != null) {   // The model can be destroyed by it's itemTreeManager:
      // the save action close the Item Tree window
      item = getSelectedItem()
      getModel()!!.setActorEnabled(VItemTree.CMD_QUIT, true)
      if (item != null) {
        getModel()!!.setActorEnabled(VItemTree.CMD_ADD, getModel()!!.isInsertMode)
        getModel()!!.setActorEnabled(VItemTree.CMD_REMOVE, getModel()!!.isInsertMode && item.id != -1)
        getModel()!!.setActorEnabled(VItemTree.CMD_EDIT, getModel()!!.isInsertMode && item.id != -1)
        getModel()!!.setActorEnabled(
          VItemTree.CMD_LOCALISE,
          getModel()!!.isInsertMode && getModel()!!.isLocalised && item.id != -1
        )
        getModel()!!.setActorEnabled(VItemTree.CMD_SELECT, !getModel()!!.isNoEdit() && item.id != -1)
        getModel()!!.setActorEnabled(VItemTree.CMD_DEFAULT, getModel()!!.isMultiSelectionDefaultValue() && item.id != -1)
        if (isLeaf()) {
          getModel()!!.setActorEnabled(VItemTree.CMD_FOLD, false)
          getModel()!!.setActorEnabled(VItemTree.CMD_UNFOLD, false)
        } else {
          if (tree.isExpanded(tree.selectedItem)) {
            getModel()!!.setActorEnabled(VItemTree.CMD_FOLD, true)
            getModel()!!.setActorEnabled(VItemTree.CMD_UNFOLD, false)
          } else {
            getModel()!!.setActorEnabled(VItemTree.CMD_FOLD, false)
            getModel()!!.setActorEnabled(VItemTree.CMD_UNFOLD, true)
          }
        }
      } else {
        getModel()!!.setActorEnabled(VItemTree.CMD_FOLD, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_UNFOLD, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_ADD, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_REMOVE, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_EDIT, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_LOCALISE, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_SELECT, false)
        getModel()!!.setActorEnabled(VItemTree.CMD_DEFAULT, false)
      }
      getModel()!!.setActorEnabled(VItemTree.CMD_SAVE, getModel()!!.isChanged)
    }
  }

  override fun run() {
    isVisible = true
    tree.focus()
    tree.expand(tree.getItem(-1))
  }

  /**
   * Returns the selected item.
   * @return The selected item.
   */
  fun getSelectedItem(): Item? {
    return if (tree.selectedItem != null) {
      tree.getITEM(tree.selectedItem)
    } else null
  }

  /**
   * Returns true if the selected item is leaf
   */
  fun isLeaf(): Boolean {
    return !tree.dataProvider.hasChildren(tree.selectedItem)
  }

  // --------------------------------------------------------------------
  // ACCESSORS
  // --------------------------------------------------------------------
  override fun getModel(): VItemTree? = super.getModel() as VItemTree?

  // --------------------------------------------------
  // UItemTree IMPLEMENTATION
  // --------------------------------------------------
  override fun getTree(): UItemTree.UTreeComponent = tree
  // --------------------------------------------------
  // LISTENERS
  // --------------------------------------------------
  /**
   * The `ItemClickHandler` is the item tree implementation
   * of the [ItemClickListener].
   */
  private inner class ItemClickHandler : ComponentEventListener<ItemClickEvent<TreeNode>> {
    override fun onComponentEvent(event: ItemClickEvent<TreeNode>) {
      if (event.clickCount == 2) {
        try {
          setSelectedItem()
        } catch (e: VException) {
          throw VRuntimeException(e.message, e)
        }
      } else {
        setTree()
      }
    }
  }

  /**
   * The `CollapseHandler` is the item tree implementation
   * of the [CollapseListener].
   */
  /*private inner class CollapseHandler : ComponentEventListener<CollapseEvent<TreeNode, TreeGrid<TreeNode>>> {
    override fun onComponentEvent(event: CollapseEvent<TreeNode, TreeGrid<TreeNode>>) {
      tree.valueChanged()
    }
  }*/

  /**
   * The `ExpandHandler` is the item tree implementation
   * of the [ExpandListener].
   */
  /*private inner class ExpandHandler : ComponentEventListener<ExpandEvent<TreeNode, TreeGrid<TreeNode>>> {
    override fun onComponentEvent(event: ExpandEvent<TreeNode, TreeGrid<TreeNode>>) {
      tree.valueChanged()
    }
  }*/
}
